### DISCUSSION TITLE
Set up of the asio4all-driver

### DISCUSSION THREAD

**Discussion by momusik** [jamulussoftware/jamulus#2163] [Technical Support & Help]
Hello,
I am looking for a german speaking person who can help me with the following informations I get when I try to do the set up of the asio4all-driver. Can anybody help me, what I need to do now?

Hallo,
ich habe das Programm Jamulus und den asio4all-Treiber installiert. Wenn ich die Einstellungen im Asio-Treiber vornehmen will, bekomme ich die unten stehende Fehlermeldung. Was bedeutet das? Kann mir jemand helfen? 

![grafik](https://user-images.githubusercontent.com/95779002/145284134-ea7560b6-7a58-4bb7-b41c-9d6bfc001d2b.png)
![grafik](https://user-images.githubusercontent.com/95779002/145284836-b3ef4786-e02d-4bd0-8743-fdbbaeae590e.png)

**Comment by ann0see** [comment]
Starte den PC neu und wähle dann ein Gerät aus, das 48kHz Sample Rate unterstützt. Das müsste dir die neue ASIO4ALL Version anzeigen. 
 
 ASIO4ALL ist nicht wirklich kompatibel mit vielen integrierten Soundkarten und bereitet meiner Meinung nach viel zu viel Kopfzerbrechen und Zeit.

Das Geld das du glaubst zu sparen wenn du ASIO4ALL ohne externe Hardware nutzt statt Interface/USB Mikro zu kaufen was sogar mit ASIO4ALL eigentlich relativ gut funktionieren müsste, gibst du in Form von Zeit und Frustration aus (meiner Erfahrung nach aus dem Chor).

**Comment by ann0see** [comment]
@gilgongo how can we make this error message clearer? 

@momusik we’d like to get a few more insights how this error message can be made clearer to improve the - known - user experience issues with ASIO4ALL.

**Reply by ann0see** [reply]
I think the please choose another device message should not be where it is —> improve the error listing.

**Reply by gilgongo** [reply]
@ann0see The classic answer to making an error message better is to design the system so that the message isn't needed any more :-) 
But what's the circumstance here (and can you point me to the current English)?

**Reply by ann0see** [reply]
No usable ASIO Device driver found would be the message. 

 The second part must be raised somewhere in Windows/Sound.cpp with the sample rate mismatch.

**Reply by ann0see** [reply]
Basically this message comes up if there’s no compatible device. What happens under the hood is:

check each available device:
 is it compatible, if no: save the error message
 if yes, use this device (or do something else I don’t remember)

if no compatible device was found:
Output all the error messages associated with all devices (which is only one in this case). By showing this error message basically we say "Ah yeah every device you have is incompatible. Please have a look at the error messages associated with every device (maybe it helps you debugging):
* ASIO4ALL: you chose a device not supporting 48kHz"

The problem is that the introduction is much too wordy.

Maybe the edge case with only one device and one error message should be handled differently.

**Reply by gilgongo** [reply]
I wonder if the confusion might also come from the fact that the user might not be expecting these messages? If so, then this lends more weight to the idea of the "setup wizard" you had I think? So if they start on a journey of "Test your devices for compatibility with Jamulus" and it explains what it'll do before they get this message, then it might be easier to deal with? 

BTW this might be a tangent, but is this issue now solved? https://github.com/jamulussoftware/jamulus/issues/1338

**Reply by ann0see** [reply]
Yeah. The setup wizard doesn’t exist (yet). No, the issue is not yet solved due to the macOS issues in the linked PR. And yes, I think both are tangent

**Reply by gilgongo** [reply]
OK. So when you say "The problem is that the introduction is much too wordy" that sounds like a good place to start. Can you link me to the English wording in the source?

**Reply by ann0see** [reply]
Basically the error message is created here: https://github.com/jamulussoftware/jamulus/blob/ccdb3ddad5861a860b1ee5de5953da02c8289895/src/soundbase.cpp#L166-L197

**Reply by ann0see** [reply]
We could maybe remove the whole 

```
                                     "</b><br><br>" + tr ( "These are all the available drivers with error messages:" ) + "<ul>"; 
```

if there's only one device

```
if ( !vsErrorList.isEmpty() )
       {
           // create error message with all details
           QString sErrorMessage = "<b>" + QString ( tr ( "No usable %1 audio device found." ) ).arg ( strSystemDriverTechniqueName ) + "</b>";
          if ( lNumDevs > 1 )
          {   // only output multiple messages if more then one are necessary.
              sErrorMessage += "<br><br>" + tr ( "These are all the available drivers with error messages:" ) + "<ul>";
              for ( int i = 0; i < lNumDevs; i++ )
              {
                  sErrorMessage += "<li><b>" + GetDeviceName ( i ) + "</b>: " + vsErrorList[i] + "</li>";
              }
              sErrorMessage += "</ul>";
          } 
          else
          {
              sErrorMessage += "<b>" + GetDeviceName ( 0 ) + "</b>: " + vsErrorList[0] ;
          }



#ifdef _WIN32
           // to be able to access the ASIO driver setup for changing, e.g., the sample rate, we
           // offer the user under Windows that we open the driver setups of all registered
           // ASIO drivers
           sErrorMessage += "<br/>" + tr ( "You might be able to fix this error by changing settings of your ASIO driver. Do you want to open these settings now?" );

           if ( QMessageBox::Yes == QMessageBox::information ( nullptr, APP_NAME, sErrorMessage, QMessageBox::Yes | QMessageBox::No ) )
           {
               LoadAndInitializeFirstValidDriver ( true );
           }

           sErrorMessage = QString ( tr ( "Can't start %1. Please restart %1 and check/reconfigure your audio settings." ) ).arg ( APP_NAME );
#endif

           throw CGenErr ( sErrorMessage );
       }
```

Changed a little bit more.

**Comment by dcorson-ticino-com** [comment]
Hi Mo,
Bist du ganz sicher das kein anderes Programm die Audio-Ausgang nicht schon benützt ?
Im Normalfall erlaubt Win10 nicht das mehrere Programme gleichzeitig Audio benützen.
PC neu starten und dann probieren.

**Comment by momusik** [comment]
Hallo, 
vielen Dank für die Rückmeldungen und Unterstützung.
Ein Neustart hat leider nicht geholfen.
Ich habe Jamulus installiert und den aktuellen Asio4all-Treiber (2.15 Beta2, Englisch)
Ich verwende folgende Geräte: Behringer Xenyx 302 USB Mixer, Ultravoice XM8500 Mikrofon, HPM1000 Kopfhörer
Ich habe das Interface (USB-Anschluss) mit meinem Laptop verbunden.
Das Mikrofon habe ich in das Audio Interface gesteckt.
Den Kopfhörer habe ich in die Buchse am Laptop gesteckt. Es wurde als Lautsprecher erkannt:
![grafik](https://user-images.githubusercontent.com/95779002/145382018-3a02fba7-89b0-46d3-94d2-9fb907f61954.png)
Folgende Einstellungen habe ich vorgenommen unter Windows Soundeinstellungen:
![grafik](https://user-images.githubusercontent.com/95779002/145382587-0b51cecb-2bcd-4331-9c96-80366736cafd.png)
![grafik](https://user-images.githubusercontent.com/95779002/145382647-62b3309d-4a02-4aeb-9fcd-92faacc462d8.png)
![grafik](https://user-images.githubusercontent.com/95779002/145383147-df9d030b-ef49-4f8b-975d-1a18c93c3ef1.png)
![grafik](https://user-images.githubusercontent.com/95779002/145383019-b314260d-ee45-4579-8c64-8e3b34982d60.png)
![grafik](https://user-images.githubusercontent.com/95779002/145382711-8446e470-759a-42ee-a6a7-0b0657dc84fa.png)

Die Meldung, die ich unter Jamulus / Einstellungen / Erweiterte Einstellungen / Audio -/ Netzwerkeinstellungen / Asio-Geräte-Einstellungen
finde, sende ich extra.

**Comment by momusik** [comment]
![grafik](https://user-images.githubusercontent.com/95779002/145386316-ac12e87c-9a86-424f-8cf9-75f83067abdf.png)

Weitere Einstellungen: 
Audiokanäle: Mono-In / Stereo-Out
Audioqualität: Normal
Puffergröße: 2,67 ms (64)
Netzwerkpuffer: Auto
Aktiviere kleinen Netzwerkpuffer (aktiviert)

**Reply by ann0see** [reply]
Muss es wirklich der Output vom PC sein? Häufig gibt es Probleme, wenn man 2 verschiedene Geräte nutzen möchte. Wenn der Mixer einen Audio Ausgang für Große Klinke hat, müsste es relativ günstige Adapter geben. Dann kannst du ein und Ausgang über den Mixer nutzen und die integrierte Karte ausstellen.

**Comment by momusik** [comment]
Ist damit die Variante B (siehe unten) gemeint?
Hört man dabei denn auch die anderen Teilnehmer bei Jamulus?
Ich dachte, wenn man die Variante nimmt, hört man nur sich selbst, also den Ton vom Mikrofon, das in das Gerät eingesteckt ist.

![grafik](https://user-images.githubusercontent.com/95779002/145415779-ae0213ff-d376-4d39-8b3c-efd5180114a4.png)

**Reply by ann0see** [reply]
Genau. Variante B. Die Frage ist natürlich, ob der Mixer es erlaubt Computersound auszugeben.
